### PR TITLE
IBP-3777-Printing label Obs and Adding Lot/Transaction 

### PR DESCRIPTION
**Pull request by cuenyad**
Please, review.

Include:
https://github.com/IntegratedBreedingPlatform/BMSAPI/pull/381
https://github.com/IntegratedBreedingPlatform/Fieldbook/pull/903
https://github.com/IntegratedBreedingPlatform/Workbench/pull/364

Regards, Diego

### CODE REVIEW

File: src/main/java/org/generationcp/middleware/dao/dms/ExperimentDao.java [hunk 1]
```diff
@@ -945,6 +946,54 @@ public ExperimentModel save(final ExperimentModel entity) {
 		}
 	}
 
+	public List<Map<String, Long>> getInstanceInformation(final Integer datasetId, final Integer studyId) {
```
Comment (nahuel-soldevilla) [inline src/main/java/org/generationcp/middleware/dao/dms/ExperimentDao.java:949]: Maybe a DTO (`aliasToBean`) would be better

File: src/main/java/org/generationcp/middleware/api/inventory/study/StudyTransactionsDto.java [hunk 2]
```diff
@@ -12,12 +12,12 @@
 public class StudyTransactionsDto extends TransactionDto {
 
 	public StudyTransactionsDto(final Integer transactionId, final String createdByUsername, final String transactionType,
-		final Double amount, final String notes, final Date createdDate, final Integer lotId, final String lotUUID,
+		final Double amount, final Double availableBalance, final String notes, final Date createdDate, final Integer lotId, final String lotUUID,
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/api/inventory/study/StudyTransactionsDto.java:15]: rename to lotAvailableBalance

File: src/main/java/org/generationcp/middleware/domain/inventory/manager/TransactionDto.java [hunk 3]
```diff
@@ -28,7 +28,7 @@ public TransactionDto() {
 	}
 
 	public TransactionDto(
-		final Integer transactionId, final String createdByUsername, final String transactionType, final Double amount,
+		final Integer transactionId, final String createdByUsername, final String transactionType, final Double amount, final Double availableBalance,
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/domain/inventory/manager/TransactionDto.java:31]: rename to lotAvailableBalance

File: src/main/java/org/generationcp/middleware/service/api/dataset/DatasetService.java [hunk 4]
```diff
@@ -399,4 +399,13 @@ FilteredPhenotypesInstancesCountDTO countFilteredInstancesAndPhenotypes(
...
 	Table<Integer, Integer, Integer> getTrialNumberPlotNumberObservationUnitIdTable(final Integer datasetId, final Set<Integer> instanceNumbers, final Set<Integer> plotNumbers);
+
+	/**
+	 * Given a dataset Id and Study Id, it will retrieve how many observations, number of repetitions, and entries it has associated.
+	 *
+	 * @param datasetId
+	 * @param studyId
+	 * @return List<InstanceInfomationDTO>
+	 */
+	List<InstanceInfomationDTO> getInformationInstance(final Integer datasetId, final Integer studyId);
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/service/api/dataset/DatasetService.java:410]: remove finals

File: src/main/java/org/generationcp/middleware/service/api/dataset/InstanceInfomationDTO.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
+package org.generationcp.middleware.service.api.dataset;
+
+import org.pojomatic.Pojomatic;
+import org.pojomatic.annotations.AutoProperty;
+
+@AutoProperty
+public class InstanceInfomationDTO {
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/service/api/dataset/InstanceInfomationDTO.java:7]: InstanceMetadataDTO or InstanceDetailsDTO sounds better to me

File: src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java [hunk 6]
```diff
@@ -1312,6 +1313,13 @@ public void setWorkbenchDataManager(final WorkbenchDataManager workbenchDataMana
 		return this.daoFactory.getExperimentDao().countObservationsPerInstance(datasetId);
 	}
 
+
+	@Override
+	public List<InstanceInfomationDTO> getInformationInstance(final Integer datasetId, final Integer studyId) {
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/service/impl/dataset/DatasetServiceImpl.java:1318]: rename by getInstanceInformation or (getInstanceDetails or getInstanceMetadata if the class name is modified)

File: src/main/java/org/generationcp/middleware/dao/ims/TransactionDAO.java [hunk 7]
```diff
@@ -380,6 +380,9 @@ private void createTransactionRow(final List<TransactionReportRow> transactionRe
 		+ "' WHEN trntype = " + TransactionType.ADJUSTMENT.getId() + "  THEN '" + TransactionType.ADJUSTMENT.getValue()
 		+ "' END) AS transactionType,"//
 		+ "    tr.trnqty AS amount,"//
+		+ "(SELECT SUM(CASE WHEN transaction.trnstat = " + TransactionStatus.CONFIRMED.getIntValue()
```
Comment (clarysabel) [inline src/main/java/org/generationcp/middleware/dao/ims/TransactionDAO.java:383]: there is any other way to calculate this outside this query? we were avoiding including the calculation of aggregated data for the lots in this query.

**Pull request merged**